dfc/tests/integration_test.go

```diff
@@ -0,0 +1,301 @@
+package dfc_test
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/NVIDIA/dfcpub/dfc"
+	"github.com/NVIDIA/dfcpub/pkg/client"
+)
+
+type repFile struct {
+	repetitions int
+	filename    string
+}
+
+type metadata struct {
+	t                *testing.T
+	smap             dfc.Smap
+	delay            time.Duration
+	semaphore        chan struct{}
+	repFilenameCh    chan repFile
+	wg               *sync.WaitGroup
+	bucket           string
+	targetDirectURL  string
+	sid              string
+	origNumTargets   int
+	num              int
+	numGetsEachFile  int
+	numGetErrsBefore uint64
+	numGetErrsAfter  uint64
+	getsCompleted    uint64
+	reregistered     uint64
+}
+
+// Intended for a deployment with multiple targets
+// 1. Unregister target T
+// 2. Create local bucket
+// 3. PUT large amount of objects into the local bucket
+// 4. GET the objects while simultaneously re-registering the target T
+func TestGetAndReRegisterInParallel(t *testing.T) {
+	var (
+		err error
+		num = 20000
+		m   = metadata{
+			t:               t,
+			delay:           15 * time.Second,
+			num:             num,
+			numGetsEachFile: 5,
+			repFilenameCh:   make(chan repFile, num),
+			semaphore:       make(chan struct{}, 10), // 10 concurrent GET requests at a time
+			wg:              &sync.WaitGroup{},
+		}
+		// Currently, a small percentage of GET errors can be reasonably expected as a result of this test.
+		// With the current design of dfc, there is exists a brief period in which the cluster map is synced to
+		// all nodes in the cluster during re-registering. During this period, errors can occur.
+		maxNumGetErrs = uint64(m.num * m.numGetsEachFile / 10) // 10 % of GET requests
+		filenameCh    = make(chan string, m.num)
+		errch         = make(chan error, 100)
+		sgl           *dfc.SGLIO
+		filesize      = uint64(1024)
+		seed          = int64(111)
+	)
+	// Step 1.
+	m.smap, err = client.GetClusterMap(proxyurl)
+	checkFatal(err, t)
+
+	l := len(m.smap.Tmap)
+	if l < 2 {
+		t.Fatalf("Must have 2 or more targets in the cluster, have only %d", l)
+	}
+	m.origNumTargets = l
+	for m.sid = range m.smap.Tmap {
+		break
+	}
+	m.targetDirectURL = m.smap.Tmap[m.sid].DirectURL
+	err = client.UnregisterTarget(proxyurl, m.sid)
+	checkFatal(err, t)
+
+	n := len(getClusterMap(httpclient, t).Tmap)
+	tlogf("Unregistered target %s: the cluster now has %d targets\n", m.sid, n)
+
+	// Step 2.
+	m.bucket = TestLocalBucketName
+	err = client.CreateLocalBucket(proxyurl, m.bucket)
+	checkFatal(err, t)
+
+	defer func() {
+		err = client.DestroyLocalBucket(proxyurl, m.bucket)
+		checkFatal(err, t)
+	}()
+
+	if usingSG {
+		sgl = dfc.NewSGLIO(filesize)
+		defer sgl.Free()
+	}
+
+	// Step 3.
+	tlogf("Putting %d files into bucket %s...\n", num, m.bucket)
+	putRandomFiles(0, seed, filesize, num, m.bucket, t, nil, errch, filenameCh, SmokeDir, SmokeStr, "", true, sgl)
```

Comment:
use m.num?

Reply:
`num` is defined at the top of the function as well, so `num` and `m.num` are the same here.